073-extend-feature-gates-to-all-operators.md

```diff
@@ -0,0 +1,116 @@
+# Extend Feature Gates to all Strimzi operators
+
+Strimzi introduced _Feature Gates_ in [Proposal no. 22](https://github.com/strimzi/proposals/blob/main/022-feature-gates.md).
+Feature gates allow for the enabling or disabling of certain features that change the behavior of Strimzi.
+They also have a maturity model that allows the feature gates to progress as the features they introduce mature.
+For more details, please check out the [original proposal](https://github.com/strimzi/proposals/blob/main/022-feature-gates.md).
+The original proposal introduced feature gates initially only to the Strimzi Cluster Operator.
+This proposal provides the plan for extending them to all Strimzi operators.
+
+## Current situation
+
+Currently, the feature gates are used only in the Cluster Operator.
+The feature gates are configured using `STRIMZI_FEATURE_GATES` environment variable where the individual gates can be enabled or disabled.
+The configured feature gates are used across the Cluster Operator code.
+Also the code implementing them is part of the `cluster-operator` module.
+The feature gates do not extend to the other operators (User and Topic Operator).
+
+## Motivation
+
+Although the feature gates currently don't apply to the other operators, some of them still affect them:
+* For example, the `UnidirectionalTopicOperator` feature gate controlled the mode in which the Topic Operator was running.
+  But it was using different environment variables to configure the right mode Topic Operator and not the feature gates configuration.
+* In the future, the `UseServerSideApply` feature gate (see [Proposal no. 52](https://github.com/strimzi/proposals/blob/main/052-k8s-server-side-apply.md) for more details) will apply to at least the User Operator.
+
+Additional feature gates that apply to the other operators might be added in the future.
+To have a unified way how to manage such feature gates, this proposal provides the plan for extending the feature gates to the User and Topic Operators as well.
+
+## Proposal
+
+The feature gates and the `STRIMZI_FEATURE_GATES` environment variable will keep its current format.
+The support for the `STRIMZI_FEATURE_GATES` environment variable will be added to the User and Topic Operators.
+The same feature gates will be defined and handled by every operator.
+However, if the operator doesn't have any use for the feature gate, it can simply ignore it in its code.
```

Comment:
In the implementation can we log (e.g. at INFO) the feature gates which are, and are not, understood/being consumed by an operator?

Reply:
To be honest, I did not plan for that. I don't think it is completely trivial ... I think it leads to a small _spiral of complexity_ ...
* I do not think you would want to have a dedicated code in each operator with hardcoded log message as that might fail to get updated etc.
* So you probably need to have some flag in the Feature Gate itself ... `supportsTopicOperator`, `supportsClusterOperator`, etc.
* But if you have that flag, should you really use it just for logging? Maybe you can use it earlier, not pass the irrelevant feature gates down the stream (e.g. not update the running UO when a feature gate that is not used by UO is not enabled), and save an unnecessary rolling update.
* But if you do that, what about the standalone UO and TO ... should it reject the feature gates it does not use as if it was an invalid feature gate? Or what should it do?

So I wonder ... is it really worth the effort and the complexity? Does it make sense to stop only with the log message if you already have the information? 

I think it is not worth it. We can for example use the docs to explain which components are affected by the feature gate and I think that would be sufficient. But obviously, it is not like the implementation would take weeks ... so if people think this would be worth it, I can update the proposal.

Reply:
@tombentley So, what should we do about this? Everyone seems to have approved it without this being incorporated. But I can still try to incorporate it and restart the reviews if you want to have it in.

Reply:
My concern was it just opens the door to users thinking they've configured something, but not actually, because the operator they're configuring it on doesn't honour it. These are the sorts of dumb mistakes that are among the most frustrating.

I wasn't thinking of a flag-based mechanism that you propose with `supportsTopicOperator` etc. Instead I was thinking about encapsulating the flags so that an operator can only get the enabled flags by passing in the set of flats it supports. So you have some class which does the initial parsing of the env var into some `FeatureFlags` object (which can reject completely unknown flags), but that doesn't directly expose methods for getting the enabled flags. Instead it exposes a method like `operatorScopedFlags(Set<String> operatorFlags)` which accepts the set of flags understood by that operator, and returns the subset of enabled flags which the operator can then use. That allows the `operatorScopedFlags()` to log about flags which we passed by the user, but not requested/consumed by the operator, but without having to model that within `common`. 

Obviously this doesn't allow the rolling avoidance feature you mention, but I guess no one thinks that's a big enough problem (or else they wouldn't have voted to accept this proposal).

All that said, if no one else thinks this is a problem worth solving, or if that just a lot of refactoring, then I'm happy if you proceed with out it. It's something we could revisit if it turned out to be a problem in practice.

Reply:
Ok, I guess given the number of approvals, we should keep it as it is and we can extend it if we see any issues. Thanks.